src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/OnTypeFormattingHandler.cs

```diff
@@ -0,0 +1,174 @@
+            {
+                return EmptyEdits;
+            }
+
+            // Switch to a background thread.
+            await TaskScheduler.Default;
+
+            var projectionResult = await _projectionProvider.GetProjectionAsync(documentSnapshot, request.Position, cancellationToken).ConfigureAwait(false);
+            if (projectionResult == null || projectionResult.LanguageKind != RazorLanguageKind.Html)
+            {
+                return EmptyEdits;
+            }
+
+            if (request.Options.OtherOptions == null)
+            {
+                request.Options.OtherOptions = new Dictionary<string, object>();
+            }
+            request.Options.OtherOptions[LanguageServerConstants.ExpectsCursorPlaceholderKey] = true;
+
+            var formattingParams = new DocumentOnTypeFormattingParams()
+            {
+                Character = request.Character,
+                Options = request.Options,
+                Position = projectionResult.Position,
+                TextDocument = new TextDocumentIdentifier() { Uri = projectionResult.Uri }
+            };
+
+            var serverKind = projectionResult.LanguageKind == RazorLanguageKind.CSharp ? LanguageServerKind.CSharp : LanguageServerKind.Html;
+            var edits = await _requestInvoker.RequestServerAsync<DocumentOnTypeFormattingParams, TextEdit[]>(
+                Methods.TextDocumentCompletionName,
+                serverKind,
+                formattingParams,
+                cancellationToken).ConfigureAwait(false);
+
+            if (edits == null)
+            {
+                return EmptyEdits;
+            }
+
+            var mappedEdits = new List<TextEdit>();
+            foreach (var edit in edits)
+            {
+                if (edit.Range == null || edit.NewText == null)
+                {
+                    // Sometimes the HTML language server returns invalid edits like these. We should just ignore those.
+                    continue;
+                }
+
+                var mappingResult = await _documentMappingProvider.RazorMapToDocumentRangeAsync(projectionResult.LanguageKind, request.TextDocument.Uri, edit.Range, cancellationToken).ConfigureAwait(false);
+
+                if (mappingResult == null || mappingResult.HostDocumentVersion != documentSnapshot.Version)
+                {
+                    // Couldn't remap the edits properly. Discard this request.
+                    return EmptyEdits;
+                }
+
+                var mappedEdit = new TextEdit()
+                {
+                    NewText = edit.NewText,
+                    Range = mappingResult.Range
+                };
+                mappedEdits.Add(mappedEdit);
+            }
+
+            if (mappedEdits.Count == 0)
+            {
+                return EmptyEdits;
+            }
+
+            await _joinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
A test fails on this line,
`Microsoft.VisualStudio.Threading.JoinableTaskContextException : An attempt to switch to the main thread failed to reach the expected thread. Was the JoinableTaskContext initialized on the wrong thread or with a SynchronizationContext whose Post method does not execute its delegate on the main thread?`

In this method I am,
- Switching to Main Thread to get the document snapshot
- Switching to a background thread to do all sorts of projection, querying the html server, mapping etc
- Switching back to the main thread to apply edits

Looks like it doesn't want me to switch back to the main thread after explicitly switching away from it the first time. If I remove `await TaskScheduler.Default;` on line 95 then the test passes. Thoughts?

Reply:
Interesting... It's probably because the"main thread" that we fake create in our test doesn't guarantee that any thread can hop back onto it given its managed by the thread pool. I'd throw the await Task.Default into a virtual method and at test time leave yourself on the main thread when attempting to move onto the BG thread